programs/native/template_signer/src/lib.rs

```diff
@@ -0,0 +1,579 @@
+    /// mask for the message. 0 bits are to be filled out by the request.
+    mask: Vec<u8>,
+    /// challenge to validate the key
+    challenge: Vec<u8>,
+    /// signature of the challenge
+    sig: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Default)]
+struct Request {
+    /// (request.msg & template.mask) == (template.msg & template.mask)
+    msg: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Default)]
+struct SigningContract {
+    /// Template that will be filled by the signer upon request
+    template: Template,
+    /// The identity that can claim the escrow and `guarantee` if the template is signed.
+    signer: Pubkey,
+    /// The signing `guarantee`.
+    guarantee: u64,
+    /// The tokens in escrow to be released to the signer if they sign the template.
+    escrow: u64,
+    /// Request to sign must be made before template_timeout.
+    template_timeout: u64,
+    /// Once request to sign has been made the signer needs to submit the signature before this
+    /// timeout.
+    requested_timeout: u64,
+    /// Once the template has been signed and `signed_timeout` expires the `guarantee` and
+    /// escrow is claimable by the signer.
+    signed_timeout: u64,
+    /// Current owner identity. This is the identity that can request a message to be signed.
+    owner: Pubkey,
+    /// current state
+    state: State,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+enum State {
+    Created,
+    EscrowFilled,
+    Requested {
+        pending: Request,
+    },
+    Signed {
+        pending: Request,
+        sig: Vec<u8>,
+    },
+    ConcurrentSignature {
+        pending: Option<Request>,
+        concurrent: Vec<u8>,
+        sig: Vec<u8>,
+    },
+    Claimed,
+}
+impl Default for State {
+    fn default() -> Self {
+        State::Created
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+enum Message {
+    /// Message to initalize the template.
+    /// SigningContract.state must be `State::Created`
+    /// SigningContract.escrow must be 0
+    /// * account[0] - must be signed, current owner
+    /// * account[1] - the SigningContract
+    Init(SigningContract),
+    /// * u64 - Tokens transfered to the SigningContract.
+    /// * account[0] - must be signed, current owner
+    /// * account[1] - the SigningContract
+    /// * account[2] - the new owner
+    ///
+    /// This should be executed in an atomic transaction with
+    /// Instruction 0, System::Move tokens from new owner to current owner
+    /// Instruction 1, Signer::InitEscrow, deposit the `escrow_token` amount into the
+    /// SigningContract
+    InitEscrow(u64),
+
+    /// * account[0] - must be signed, current owner
+    /// * account[1] - the SigningContract
+    /// * account[2] - the new owner
+    ///
+    /// This should be executed in an atomic transaction with
+    /// Instruction 0, System::Move tokens from new owner to current owner
```

Comment:
Should the SigningContract itself hold the tokens?

Reply:
@mvines yea, but i want the coordination of transferring ownership for tokens to be implemented by another program.